sklearn/tree/_criterion.pyx

```diff
@@ -1238,18 +1238,22 @@ cdef class MAE(RegressionCriterion):
         cdef SIZE_t* samples = self.samples
         cdef SIZE_t i, p, k
         cdef DOUBLE_t y_ik
-        cdef DOUBLE_t w_y_ik
-
-        cdef double impurity = 0.0
+        cdef DOUBLE_t w = 1.0
+        cdef DOUBLE_t impurity = 0.0
 
         for k in range(self.n_outputs):
             for p in range(self.start, self.end):
                 i = samples[p]
 
                 y_ik = y[i * self.y_stride + k]
 
-                impurity += <double> fabs((<double> y_ik) - <double> self.node_medians[k])
-        return impurity / (self.weighted_n_node_samples * self.n_outputs)
+                if sample_weight != NULL:
+                    w = sample_weight[i]
+
+                impurity += fabs(y_ik - self.node_medians[k]) * w
+
+        return <double>(impurity / (self.weighted_n_node_samples *
```

Comment:
Note that impurity has been changed to a type DOUBLE_t so that the calculation within the loop does not require any implicit conversion since y_ik, self.node_medians and w are all now DOUBLE_t.
 
I think all the other variables would be converted to DOUBLE_t (not double) as this will likely be the "bigger" datatype...?  I'd agree with you on the whole though, after doing this end calculation with all variables implicitly DOUBLE_T, the compiler will do one final implicit conversion to <double>.  So explicit cast not required.  I'll remove explicit cast.